src/id3manager/formats/abc.py

```diff
@@ -0,0 +1,23 @@
+import abc
+import typing as t
+
+import mutagen.id3 as id3
+
+
+class MetadataReader(abc.ABC):
+    @abc.abstractmethod
+    def read(self) -> t.List[id3.Frame]:
+        """Deserialize ID3 frames from a stream of bytes."""
+
+
+class MetadataWriter(abc.ABC):
+    @abc.abstractmethod
+    def write(self, frames: t.List[id3.Frame]) -> None:
+        """Serialize ID3 frames into a stream of bytes."""
+
+
+class IOWrapper:
```

Comment:
I don't want to be an asshole, I really do appreciate all the hard work you put into designing this, but I want to discuss the alternative route before accepting this suggestion.

I kinda feel that `IOWrapper` complicates a bit `id3manager` internals (e.g. it introduces `get_reader`/`get_writer` helpers). While dealing with a `fileobj` makes perfect sense in terms of being in control of efficient implementation (i.e. reading/writing in chunks) I wonder is this something worth investing into?

I mean we don't expect the data to be too big, and most libraries, such as `tomlkit` or `json` (not saying our own text metadata formatter), actually do `f.read()`/`f.write()` under the hood. I.e. it's not like we're taking advantage of dealing with `fileobj`.

So the alternative solution would be to make readers/writers to receive and return bytes (or maybe even strings), and let the application layer to deal with fileobj. I'm by no means enforcing this, I'm rather wondering what your thoughts on it?

I do see benefits of dealing with `fileobj` since it provides opportunities to write memory efficient formatters, yet it does look a bit awkward that `MetadataWriter` receives frames and returns nothing. But maybe I'm overthinking it. :thinking:

Reply:
Hmm, I don't think it's necessarily `fileobj` which is a problem here, but rather the unclear relationship between `IOWrapper` and `MetadataReader`/`MetadataWriter`. The wrapper's only purpose is to have a common way to accept and store a file-like object, so that reader/writer implementation have access to it. That only saves us a couple of lines... and it's only needed if readers and writers are _classes_. We don't have to go that way and we could just use regular functions.

WDYT about something like https://github.com/ikalnytskyi/id3manager/pull/11/files instead?

Reply:
> The wrapper's only purpose is to have a common way to accept and store a file-like object, so that reader/writer implementations have access to it.

But `fileobj` is kinda part of the interface since the caller passes it to `__init__()`. Shall we maybe just add the `fileobj` to the interface and call it a day?

> WDYT about something like https://github.com/ikalnytskyi/id3manager/pull/11/files?

And now I'm not happy about formatters not having an explicit interface. :(

---

WDYT about slight modification of your original suggestion? Something like this:

```python
class MetadataFormatter(abc.ABC):
   name = "toml"

   def read(self, fileobj) -> t.List[Frame]:
      ...

   def write(self, fileobj, frames) -> None:
      ...
``` 

* `fileobj` is not shared between `read()`/`write()` because it's not part of the state
* there's no need to maintain the `Format` dataclass
* we enforce people to always implement both read and write methods because otherwise `edit` won't work

Reply:
Yeah, that makes sense! I think that's a good compromise. There is still a possibility to pass a file object created in the wrong mode (read vs write), but in that case an exception will be raised at runtime. Let me try this.